homeassistant/components/jellyfin/media_source.py

```diff
@@ -0,0 +1,321 @@
+        """Return all tracks of a single album as browsable media sources."""
+        tracks = await self._get_children(album_id, ITEM_TYPE_AUDIO)
+        tracks = sorted(tracks, key=lambda k: k[ITEM_KEY_INDEX_NUMBER])  # type: ignore[no-any-return]
+        return [self._build_track(track) for track in tracks]
+
+    def _build_track(self, track: dict[str, Any]) -> BrowseMediaSource:
+        """Return a single track as a browsable media source."""
+        track_id = track[ITEM_KEY_ID]
+        track_title = track[ITEM_KEY_NAME]
+        mime_type = _media_mime_type(track)
+        thumbnail_url = self._get_thumbnail_url(track)
+
+        result = BrowseMediaSource(
+            domain=DOMAIN,
+            identifier=track_id,
+            media_class=MEDIA_CLASS_TRACK,
+            media_content_type=mime_type,
+            title=track_title,
+            can_play=True,
+            can_expand=False,
+            thumbnail=thumbnail_url,
+        )
+
+        return result
+
+    async def _get_children(
+        self, parent_id: str, item_type: str
+    ) -> list[dict[str, Any]]:
+        """Return all children for the parent_id whose item type is item_type."""
+        params = {
+            "Recursive": "true",
+            "ParentId": parent_id,
+            "IncludeItemTypes": item_type,
+        }
+        if item_type == ITEM_TYPE_AUDIO:
+            params["Fields"] = ITEM_KEY_MEDIA_SOURCES
+
+        result = await self.hass.async_add_executor_job(self.api.user_items, "", params)
+        return result["Items"]  # type: ignore[no-any-return]
+
+    def _get_thumbnail_url(self, media_item: dict[str, Any]) -> str | None:
+        """Return the URL for the primary image of a media item if available."""
+        item_id = media_item[ITEM_KEY_ID]
+        image_tags = media_item[ITEM_KEY_IMAGE_TAGS]
+        api_key = self.client.config.data["auth.token"]
+
+        if "Primary" in image_tags:
+            tag = image_tags["Primary"]
+            return (
+                f"{self.url}Items/{item_id}/Images/Primary?Tag={tag}&api_key={api_key}"
+            )
+
+        return None
+
+    def _get_stream_url(self, media_item: dict[str, Any]) -> str:
+        """Return the stream URL for a media item."""
+        media_type = media_item[ITEM_KEY_MEDIA_TYPE]
+
+        if media_type == MEDIA_TYPE_AUDIO:
+            return self._get_audio_stream_url(media_item)
+
+        raise BrowseError(f"Unsupported media type {media_type}")
+
+    def _get_audio_stream_url(self, media_item: dict[str, Any]) -> str:
+        """Return the stream URL for a music media item."""
+        item_id = media_item[ITEM_KEY_ID]
+        user_id = self.client.config.data["auth.user_id"]
+        device_id = self.client.config.data["app.device_id"]
+        api_key = self.client.config.data["auth.token"]
+
+        return f"{self.url}Audio/{item_id}/universal?UserId={user_id}&DeviceId={device_id}&api_key={api_key}&MaxStreamingBitrate={MAX_STREAMING_BITRATE}"
```

Comment:
Please use [`urllib.parse.urlencode`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode) to construct urls.

Reply:
Thanks, done. Much nicer indeed.